internal/precache/workload/pullImages.go

```
@@ -0,0 +1,200 @@
+const MaxRetries int = 5
+
+// Podman auth-file related constants
+const (
+	EnvAuthFile     string = "PULL_SECRET_PATH"
+	DefaultAuthFile string = "/var/lib/kubelet/config.json"
+)
+
+// executeCmd execute shell commands
+func executeCmd(suppressCmdOutput bool, cmd string, args ...string) (err error) {
+
+	logger := log.StandardLogger()
+	logWriter := logger.Writer()
+
+	log.Debugf("Executing %s with args %s", cmd, args)
+	execCmd := exec.Command(cmd, args...)
+
+	if suppressCmdOutput {
+		execCmd.Stdout = nil
+		execCmd.Stderr = nil
+	} else {
+		execCmd.Stdout = logWriter
+		execCmd.Stderr = logWriter
+	}
+
+	if err = execCmd.Run(); err != nil {
+		if !suppressCmdOutput {
+			log.Error(err)
+		}
+	}
+
+	_ = logWriter.Close()
+	return err
+}
+
+// CheckPodman verifies that podman is running by checking the version of podman
+func CheckPodman() bool {
+	if err := executeCmd(false, "podman", []string{"version"}...); err != nil {
+		return false
+	}
+	return true
+}
+
+// podmanImgExists reports the existence of the given image via podman CLI
+func podmanImgExists(image string) bool {
+	if err := executeCmd(true, "podman", []string{"image", "exists", image}...); err != nil {
+		return false
+	}
+	return true
+}
+
+// podmanImgPull pulls the specified image via podman CLI
+func podmanImgPull(image, authFile string) error {
+	args := []string{"pull", image}
+	if authFile != "" {
+		args = append(args, []string{"--authfile", authFile}...)
+	}
+	return executeCmd(false, "podman", args...)
+}
+
+// pullImage attempts to pull an image via podman CLI
+func pullImage(image, authFile string, progress *precache.Progress) error {
+
+	var err error
+	for i := 0; i < MaxRetries; i++ {
+		err = podmanImgPull(image, authFile)
+		if err == nil {
+			log.Infof("Successfully pulled image: %s", image)
+			break
+		} else {
+			log.Infof("Attempt %d/%d: Failed to pull %s: %v", i+1, MaxRetries, image, err)
```

Can we benefit from a tiny sleep when there needs to be retry? Apparently each pull is 3x retry by default 
https://github.com/containers/podman/issues/14359 
https://github.com/containers/common/blob/main/libimage/copier.go#L297-L300

---

I can add a sleep, however, with 15 (5*3) retries, if the pull failed, clearly there is an issue (either with the image spec not being available) or the registry. The reason I avoided sleep was to fail fast as there can be easily over 100 images that need to be precached.